TACTIC: phase 1 plan complete

lunalunaa opened the pull request:
- [x] Basic "Expr with hole" mechanism with stepwise refinement
- [x] Basic error reporting for tactics 
- [x] `ExprView` for expressions
- [x] Other minor improvements 
- [x] Isolate the tactic elaboration process

Currently, users can write code like [this](https://github.com/aya-prover/aya-dev/blob/062b2c09f8416aea33429f3388f8dbdbb41efa4a/base/src/test/resources/success/src/SimpleTac.aya), where users can create "tactic blocks" fill in holes in the "tactic head" one by one in the "tactic nodes" that follows, where these nodes can contain additional holes and be filled by nested nodes.

However, tactic blocks cannot be nested (see [this](https://github.com/aya-prover/aya-dev/blob/062b2c09f8416aea33429f3388f8dbdbb41efa4a/base/src/test/resources/failure/tactic/nested-tactic.aya)) because the holes are filled "in verbatim" from tactic node inputs. This means further tactic nodes will fill the first hole of the tactic head, and if this hole is contained in a tactic block it would be very problematic.

The result of elaborating tactic block is an expression that contains no holes (and no metas when elaborated into core term).

lunalunaa commented:
try tactics but hole

ice1000 commented on base/src/main/java/org/aya/concrete/visitor/ExprFixpoint.java:
```diff
@@ -118,4 +118,15 @@
+
+  @Override default Expr visitTac(Expr.@NotNull TacExpr tactic, P p) {
+    return new Expr.TacExpr(tactic.sourcePos(), visitTacNode(tactic.tacNode(), p));
+  }
```
Should cache

ice1000 replied:
Hmmm..

ice1000 reviewed (commented):
I think we should migrate every visitor to your exprview and nuke exprvisitors. I'll do this, stay tuned

ice1000 commented on base/src/main/java/org/aya/distill/ConcreteDistiller.java:
```diff
@@ -308,6 +310,23 @@ private Doc visitAccess(Stmt.@NotNull Accessibility accessibility, Stmt.Accessib
+
+  public @NotNull Doc tacList(@NotNull Expr.TacNode.ListExprTac tacList) {
+    return Doc.cat(
+      Doc.symbol("{"),
```
`Doc.cblock`

ice1000 replied:
I'll do it

imkiva commented on base/src/main/java/org/aya/concrete/visitor/ExprOps.java:
```diff
@@ -0,0 +1,21 @@
+    return view().initial();
+  }
+
+  @Override default Expr pre(Expr expr) {
```
It's better to note the parameter and return type with `@NotNull` or `@Nullable`

imkiva commented on base/src/main/java/org/aya/concrete/visitor/ExprView.java:
```diff
@@ -0,0 +1,124 @@
+public interface ExprView {
+  @NotNull Expr initial();
+
+  default Expr pre(Expr expr) {return expr;}
```
ditto

imkiva commented on base/src/main/java/org/aya/core/term/Term.java:
```diff
@@ -221,4 +222,29 @@ public static boolean checkSubst(@NotNull SeqLike<@NotNull Param> params, @NotNu
+        return Term.this.view();
+      }
+
+      @Override public Term pre(Term term) {
```
`TermView` and `TermOps` should also be annotated with `@NotNull` or `@Nullable`, which is addressed in my HEAD. 😄

imkiva commented on cli/src/main/java/org/aya/cli/parse/AyaProducer.java:
```diff
@@ -104,7 +105,8 @@ public AyaProducer(@NotNull Either<SourceFile, SourcePos> source, @NotNull Repor
   }
 
-  @NotNull private Remark visitRemark(AyaParser.RemarkContext remark) {
+  @NotNull
+  private Remark visitRemark(AyaParser.RemarkContext remark) {
     var sb = new StringBuilder();
```
```suggestion
  private @NotNull Remark visitRemark(AyaParser.RemarkContext remark) {
    var sb = new StringBuilder();
```

imkiva reviewed (approved):
LGTM with some nitpicks

ice1000 commented on base/src/main/java/org/aya/tyck/ExprTycker.java:
```diff
@@ -338,10 +343,108 @@ else if (iter.hasNext()) {
+      case TacNode.ListExprTac listExprTac -> {
+        var tacNodes = listExprTac.tacNodes();
+        var headNode = tacNodes.first();
+        var tailNodes = tacNodes.slice(1, tacNodes.size());
```
Why not `drop(1)`?

lunalunaa replied:
fixed

ice1000 reviewed (commented):
0. It would be good for the anonymous class in `allMetas` to be a record. Furthermore, you may use `VarConsumer` to replace it
1. You created `ExprFixpoint`. I guess it's due to the rebase. I think you should remove the file.

lunalunaa commented:
Addressing the comments:
0. Yes. I'll be changing it
1. This is probably a mistake when rebasing
2. `lastly` isn't really needed because one can override `commit`. But keeping it is also fine.

ice1000 commented:
Niubi

ice1000 commented on base/src/main/java/org/aya/tyck/ExprTycker.java:
```diff
@@ -366,19 +366,20 @@ else if (iter.hasNext()) {
         var headNode = tacNodes.first();
         var tailNodes = tacNodes.drop(1);
+        // enter into a new local state
+        localCtx = localCtx.deriveMap();
```
You can create a local of the old `localCtx` variable here, and use it later to avoid `requireNotNull`

lunalunaa replied:
This is now fixed.

Yeah, there will a lot of more tests added tomorrow.

ice1000 reviewed (commented):
Maybe you want to add some "usage of variables" test cases

lunalunaa commented:
I will also be refactoring a bit because the function is really cluttered.

lunalunaa commented:
Closing this for now because I'm moving to v2.

ice1000 commented:
@lunalunaa Maybe you still want to cherry pick some of the changes which might be useful for the future
